libwnbd/wnbd_ioctl.cpp

```diff
@@ -133,6 +137,315 @@ DWORD WnbdOpenAdapterCMDeviceInstance(PDEVINST DeviceInstance)
     return Status;
 }
 
+DWORD GetDrvDetailData(HDEVINFO AdapterDevHandle, PSP_DEVINFO_DATA DevInfoData,
+                       PSP_DRVINFO_DETAIL_DATA_A DrvDetailData)
+{
+    SP_DEVINSTALL_PARAMS_A InstallParams;
+    BOOL Cleanup = FALSE, Found = FALSE;
+    SP_DRVINFO_DATA_A DrvInfoData;
+    DrvInfoData.cbSize = sizeof DrvInfoData;
+    DWORD Status = ERROR_SUCCESS;
+
+    InstallParams.cbSize = sizeof InstallParams;
+    if (!SetupDiGetDeviceInstallParamsA(AdapterDevHandle, DevInfoData, &InstallParams)) {
+        Status = GetLastError();
+        goto exit;
+    }
+
+    InstallParams.FlagsEx |= DI_FLAGSEX_INSTALLEDDRIVER | DI_FLAGSEX_ALLOWEXCLUDEDDRVS;
+    if (!SetupDiSetDeviceInstallParamsA(AdapterDevHandle, DevInfoData, &InstallParams)) {
+        Status = GetLastError();
+        goto exit;
+    }
+
+    if (!SetupDiBuildDriverInfoList(AdapterDevHandle, DevInfoData, SPDIT_CLASSDRIVER)) {
+        Status = GetLastError();
+        goto exit;
+    }
+    Cleanup = TRUE;
+
+    for (DWORD index = 0;
+        SetupDiEnumDriverInfoA(AdapterDevHandle, DevInfoData, SPDIT_CLASSDRIVER, index, &DrvInfoData);
```

Comment:
This looked a bit odd to me at first, I didn't know why we're looping over a list of members and how to tell which is the right one.

Apparently SetupDiEnumDriverInfo lists all the drivers associated with a device. Are we sure there will be a single one, WNBD? If so, let's add an inline comment, something like: `` // SetupDiEnumDriverInfoA enumerates all the drivers associated with a device. We expect WNBD to be the only driver associated with the device``.

I'm wondering if we should throw a warning if by any chance there is more than one associated driver (maybe some improper cleanup?).

Reply:
> We shouldn't check if there are more than one associated driver. The purpose of the function is to fetch the driver information.

What if there are multiple drivers? What if we failed to uninstall a driver and a new one was added?

Reply:
This function is used to retrieve the driver information data. We only return the first element found with the aid of: `AdapterDevHandle`.
That guarantees the fact that we have a valid map between the registered device associated and the driver OEM information file.

> What if there are multiple drivers? 

Disregarding the use-case of this function.

In general If there multiple drivers installed only a single one can use the hardware node `root\wnbd` (while multiple registered devices can use that node).

Our tool only tries to uninstall the driver only if there is a a registered device using the node `root\wnbd` and does a best effort mode to remove its associated driver information file.

In my opinion we should not try to make assumption and remove all of the registered devices that are under `root\wnbd`, nor should we remove all the drivers OEM information. That could be the intend of the user.

>What if we failed to uninstall a driver and a new one was added?

Again how do you install a new one?

Reply:
> In my opinion we should not try to make assumption and remove all of the registered devices that are under root\wnbd, nor should we remove all the drivers OEM information. That could be the intend of the user.

I'd say removing all wnbd devices and drivers is the safest approach though, that's what "uninstall-driver" should do. If the user intends to keep some of those resources, that's an extremely unusual scenario. In that case, the user might as well just use devcon.exe exclusively.

Reply:
> I'd say removing all wnbd devices and drivers is the safest approach though, that's what "uninstall-driver" should do.

To be honest that sound more like a "clean"|"force" uninstall.
Uninstall should do the reverse of the install and nothing more.
It shouldn't automatically make assumption like restarting the computer if needed to force remove the current resources.
The user should not have additional installed drivers or devices if using `install-driver`.

Reply:
> To be honest that sound more like a "clean"|"force" uninstall.
Cleaning up all WNBD resources is perfectly fine IMO.

> The user should not have additional installed drivers or devices if using install-driver.
That can happen if we fail to remove the driver before attempting to install a new version. I was hitting this quite often before using our reinstall.ps1 script if there were existing mappings.